src/render/draw_particle.js

```diff
@@ -0,0 +1,154 @@
+import type ParticleBucket from '../data/bucket/particle_bucket.js';
+import type ProgramConfiguration from '../data/program_configuration.js';
+import type VertexBuffer from '../gl/vertex_buffer.js';
+import type IndexBuffer from '../gl/index_buffer.js';
+import type {UniformValues} from './uniform_binding.js';
+import type {ParticleUniformsType} from './program/particle_program.js';
+import type Tile from '../source/tile.js';
+import type {DynamicDefinesType} from './program/program_uniforms.js';
+
+export default drawParticles;
+
+type TileRenderState = {
+    programConfiguration: ProgramConfiguration,
+    program: Program<*>,
+    layoutVertexBuffer: VertexBuffer,
+    indexBuffer: IndexBuffer,
+    uniformValues: UniformValues<ParticleUniformsType>,
+    tile: Tile
+};
+
+const instanceLayout = createLayout([
+    {name: 'a_offset_and_scale', components: 4, type: 'Float32'},
+    {name: 'a_particle_color', components: 4, type: 'Float32'}
+], 4);
+
+function drawParticles(painter: Painter, sourceCache: SourceCache, layer: CircleStyleLayer, coords: Array<OverscaledTileID>) {
+    if (painter.renderPass !== 'translucent') return;
+
+    const cloudMode = layer.paint.get('particle-emitter-type') === 'cloud';
+    const opacity = layer.paint.get('particle-opacity');
+    const sortFeaturesByKey = layer.layout.get('particle-sort-key').constantOr(1) !== undefined;
+
+    if (opacity.constantOr(1) === 0 ) {
+        return;
+    }
+
+    const context = painter.context;
+    const gl = context.gl;
+
+    const depthMode = painter.depthModeForSublayer(0, DepthMode.ReadOnly);
+    // Turn off stencil testing to allow circles to be drawn across boundaries,
+    // so that large circles are not clipped to tiles
+    const stencilMode = StencilMode.disabled;
+    const colorMode = cloudMode ? ColorMode.alphaBlended : ColorMode.additiveBlended;
+
+    let particleCount = 0;
+    const particlePositions = [];
+    const tileParticleRanges = [];
+
+    for (const coord of coords) {
+        const start = particleCount;
+
+        const tile = sourceCache.getTile(coord);
+        const bucket: ?ParticleBucket<*> = (tile.getBucket(layer): any);
+
+        if (!bucket) continue;
+
+        for (const feature of bucket.features) {
+            globalSystem.addEmitter(undefined, feature.point, feature.tileId, feature.mercatorPoint, layer.paint);
+        }
+
+        globalSystem.update();
+
+        for (var emitter of globalSystem.emitters) {
+            if (!emitter.tileId.equals(bucket.tileId)) {
+                continue;
+            }
+            if (emitter.paint.get('particle-emitter-type') != layer.paint.get('particle-emitter-type')) {
+                continue;
+            }
+            for (const particle of emitter.particles) {
+                particleCount += 1;
+                particlePositions.push([
+                    emitter.location.x + emitter.zoom * particle.locationOffset.x, 
```

Comment:
Using `emitter.zoom` here is incorrect and causes issues when changing zoom levels. We would need to revisit how we store the location for the emitters and particles.